meshtal_analysis.py

```diff
@@ -44,98 +109,106 @@ def rel_err_hist(df, fname=None):
-def plot_slice(mesh, value, plane="XY", lmin=1e-15, lmax=1e-3, fname=None,
-               err=False, norm=1.0, erg=None):
+# plot slice calls extract slice
+def extract_slice(mesh, value, plane, erg=None, time=None):
     """ plots a slice through the mesh"""
-    plt.clf()
     data = mesh.data
-    # filter by energy if needed
-    if erg:
-        data = data[data["Energy"] == erg]
+    slice_obj = slice_object()
+    # filter by energy/time if needed
+    data = filter_energy_time(data, erg, time)
 
     if plane == "XZ":
-        midx = mesh.x_mids
-        midy = mesh.z_mids
-        v_mid = mesh.y_mids
+        slice_obj.slice_i = mesh.x_mids
+        slice_obj.slice_j = mesh.z_mids
+        slice_obj.axis_mids = mesh.y_mids
         v_ind = "y"
-        # ipos = "x"
-        # jpos = "z"
-        ilab = "X co-ord (cm)"
-        jlab = "Z co-ord (cm)"
+        slice_obj.i_lab = "X co-ord (cm)"
+        slice_obj.j_lab = "Z co-ord (cm)"
     elif plane == "XY":
-        midx = mesh.x_mids
-        midy = mesh.y_mids
-        v_mid = mesh.z_mids
+        slice_obj.slice_i = mesh.x_mids
+        slice_obj.slice_j = mesh.y_mids
+        slice_obj.axis_mids = mesh.z_mids
         v_ind = "z"
-        # ipos = "x"
-        # jpos = "y"
-
-        ilab = "X co-ord (cm)"
-        jlab = "Y co-ord (cm)"
+        slice_obj.i_lab = "X co-ord (cm)"
+        slice_obj.j_lab = "Y co-ord (cm)"
     elif plane == "YZ":
-        midx = mesh.y_mids
-        midy = mesh.z_mids
-        v_mid = mesh.x_mids
+        slice_obj.slice_i = mesh.y_mids
+        slice_obj.slice_j = mesh.z_mids
+        slice_obj.axis_mids = mesh.x_mids
         v_ind = "x"
-        # ipos = "y"
-        # jpos = "z"
-
-        ilab = "Y co-ord (cm)"
-        jlab = "Z co-ord (cm)"
+        slice_obj.i_lab = "Y co-ord (cm)"
+        slice_obj.j_lab = "Z co-ord (cm)"
+    else:
+        "Catch plane not recognised"
+        raise ValueError("Plane not recognised format : XZ, XY, YZ")
 
     # find closest mid point
-    value = find_nearest_mid(value, v_mid)
+    slice_obj.value = find_nearest_mid(value, slice_obj.axis_mids)
 
     # filter to just the values in the plane
-    data = data[data[v_ind] == value]
+    data = data[data[v_ind] == slice_obj.value]
 
     # now find the slice values
-    vals = np.zeros((len(midy), len(midx)))
-    err_vals = np.zeros((len(midy), len(midx)))
+    slice_obj.values = np.zeros((len(slice_obj.slice_j),
+                                 len(slice_obj.slice_i)))
+    slice_obj.errors = np.zeros((len(slice_obj.slice_j),
+                                 len(slice_obj.slice_i)))
 
     for (_, __, x, y, ____, val, rerr) in data.itertuples():
 
         x = np.float64(x)
         y = np.float64(y)
-        i, = np.where(midx == x)
-        j, = np.where(midy == y)
+        i, = np.where(slice_obj.slice_i == x)
+        j, = np.where(slice_obj.slice_j == y)
+        slice_obj.values[j, i] = val
+        slice_obj.errors[j, i] = rerr

```

Comment:
is this bit still needed once the data is filtered to just the values on the plain can we not just set slice_obj.values and errors to the dataframe columns?

Reply:
ok, leave it as is, once filtered it should be small enough of a dataframe the the loop is not too costly, but I expect there is a more efficent method